Fixes to keep OIDs across segments in sync

hlinnaka opened the pull request:
To debug issues stemming from objects having different OIDs in segments and master, I added code to print a WARNING whenever an OID is assigned to system object in segments. That should never happen, because the master should assign the OID and pass it to all the segments (although there are some exceptions where the OIDs are not kept in sync). That's the fourth patch in this series.

That WARNING revealed a few cases where OIDs are not kept in sync. ALTER TABLE didn't synchronize the OID of the type's array type (1st patch). That was missed when the patch to support array types of composite types was merged from PostgreSQL 8.3. Also, the new operator family -related DDL commands, also merged from PostgreSQL 8.3 some time ago, didn't make any effort to dispatch the commands to segments, so the segments were missing opfamily information altogether (3rd patch). Also, if CREATE OPERATOR needs to create a shell negator or commutator operator, the shell operator's OID was not synchronized (2nd patch).

Finally, this fixes the install script of the "isn" extension (5th patch). The reason it's included in this PR is that I used the isn extension to test CREATE OPERATOR FAMILY/CLASS support, because there were almost no regression tests for those. This PR adds some regression tests, but might as well fix the isn module while we're at it.

ashwinstar commented:
Looks good. Liked the WARNING to be issued while allocating OID on segment.

asimrp commented on src/backend/catalog/catalog.c:
```diff
@@ -847,6 +852,41 @@ relationId == PgFileSpaceEntryToastIndex ||
+				return false;
+
+			/* We don't currently synchronize the OIDs of these catalogs. */
+			case RewriteRelationId:
```
I'm not sure why we don't synchronize OIDs for these tables.
`CREATE VIEW` should result in identical rows added to `pg_rewrite` on master as well as segments.  I don't think GPDB supports triggers, so we can leave `pg_trigger` for now, but the same argument holds there.  Regarding `pg_amop` and `pg_amproc`, again, similar argument should hold but I don't have any experience creating a user defined access method / operator family.

hlinnaka replied:
On 22/04/16 21:02, Asim RP wrote:
> I'm not sure why we don't synchronize OIDs for these tables. `CREATE
> VIEW` should result in identical rows added to `pg_rewrite` on master
> as well as segments.  I don't think GPDB supports triggers, so we can
> leave `pg_trigger` for now, but the same argument holds there.
> Regarding `pg_amop` and `pg_amproc`, again, similar argument should
> hold but I don't have any experience creating a user defined access
> method / operator family.

I think the reason we get away without synchronizing OIDs for 
pg_rewrite, pg_amop and pg_amproc is that there are no "foreign key" 
references to those tables, and those OIDs don't appear in the Node 
structs that are dispatched from the master to segments. I'm actually 
not sure why those tables even have OIDs. At a quick glance, I don't see 
any need for them.

I'll add these musings in a comment there.

- Heikki

asimrp replied:
Thank you for the clarification, Heikki.

asimrp commented on src/backend/catalog/catalog.c:
```diff
@@ -899,27 +938,26 @@ GetNewOid(Relation relation)
+	 * RelationNeedsSynchronizedOIDs()).
+	 */
+	if (Gp_role == GP_ROLE_EXECUTE && RelationNeedsSynchronizedOIDs(relation))
+		elog(WARNING, "allocated OID %u for relation \"%s\" in segment",
```
I like this idea of issuing a warning and this is the perfect place for it.  Thanks!

asimrp commented on src/backend/commands/opclasscmds.c:
```diff
@@ -783,6 +794,12 @@ DefineOpFamily(CreateOpFamilyStmt *stmt)
+	if (Gp_role == GP_ROLE_DISPATCH)
+	{
+		stmt->newOid = opfamilyoid;
+		CdbDispatchUtilityStatement((Node *) stmt, "DefineOpFamily");
```
We should enhance `gpcheckcat` to be in sync with opfamily catalog changes.  @deepesh1081, can this be considered while your team is already working on making `gpcheckcat` better?

It may help running `gpcheckcat` after installcheck-good tests, especially when catalog changes like these are merged from upstream.

asimrp commented on src/backend/catalog/pg_operator.c:
```diff
@@ -212,7 +213,8 @@ static Oid
 				  Oid operatorNamespace,
 				  Oid leftTypeId,
-				  Oid rightTypeId)
+				  Oid rightTypeId,
```
Shell operator OIDs missed!  Quite a catch, thank you.

asimrp commented:
Typo: `s/tahat/that/` in commit message for "Give a WARNING if an object is allocated an OID in segment".

hlinnaka commented:
Added a comment to per discussion on why we don't synchronize OIDs of pg_rewrite, pg_trigger, pg_amop, fixed the typo in the comment message, and pushed. Thanks for the review, @asimrp!
